Begin implementing WebTransport network interface

achristensen07 opened the pull request:
#### 7e31db432af6660328dbd717aec83d0ac211e115
<pre>
Begin implementing WebTransport network interface
<a href="https://bugs.webkit.org/show_bug.cgi?id=278891">https://bugs.webkit.org/show_bug.cgi?id=278891</a>
<a href="https://rdar.apple.com/134977323">rdar://134977323</a>

Reviewed by Matthew Finkel.

As a first step, I start implementing the connection initialization and data sending
using h2 instead of the WebTransport protocol.  They are similar enough that they
both benefit from using the nw_connection_group_t abstraction which is an object
that can be initialized with a URL, start or receive &quot;connections&quot; and send or receive
data.  nw_connection_group_set_receive_handler needs to be called before
nw_connection_group_start so I create a NetworkTransportSession for the receive handler
before deciding if we want to discard it by calling the completion handler of
NetworkTransportSession::initialize with nullptr if there is a network failure during
initialization.  Other than that, this PR basically just adds some straightforward
abstractions and begins implementing a few of them.

WebTransport has some special needs on the server, so I made a new test server for it.
httpd doesn&apos;t really support it in a straightforward way, and even WPT&apos;s WebTransport
tests require a little custom server setup.  TestWebKitAPI::HTTPServer isn&apos;t really
the right abstraction for it, either because I will eventually need to use
nw_listener_set_new_connection_group_handler instead of nw_listener_set_new_connection_handler
to listen for quic connections.  It isn&apos;t quite the way it needs to be yet, but
I just listen for one encrypted TCP connection and wait until it receives more than 0 bytes,
which verifies that something is happening on the client.  In my next PR I&apos;ll hook up more
things and soon I&apos;ll start using nw_connection_group_t on the server too.

* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::messageSenderConnection const):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::receiveDatagram):
(WebKit::NetworkTransportSession::NetworkTransportSession): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm: Added.
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm: Added.
(TestWebKitAPI::enableWebTransport):
(TestWebKitAPI::TEST(WebTransport, Basic)):
* Tools/TestWebKitAPI/WebTransportServer.h: Added.
(TestWebKitAPI::WebTransportServer::bytesReceived const):
(TestWebKitAPI::WebTransportServer::setBytesReceived):
* Tools/TestWebKitAPI/WebTransportServer.mm: Added.
(TestWebKitAPI::connectionLoop):
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::WebTransportServer::port const):
* Tools/TestWebKitAPI/cocoa/HTTPServer.h:
* Tools/TestWebKitAPI/cocoa/WebTransportServer.h: Added.
* Tools/TestWebKitAPI/cocoa/WebTransportServer.mm: Added.

Canonical link: <a href="https://commits.webkit.org/282945@main">https://commits.webkit.org/282945@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71870630645ffe9610acf93bcbb3617748af08c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17346 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68774 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15357 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66867 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51896 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15635 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/52049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10592 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67816 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40810 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56023 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/32671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37475 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13398 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14231 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13727 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70480 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8695 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/13222 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8729 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56107 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7176 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9815 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/39927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/41006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40748 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -1964,7 +1964,6 @@
 		A14F9B762B68CA6C00AD9C56 /* WKSLinearMediaPlayer.h in Headers */ = {isa = PBXBuildFile; fileRef = A14F9B742B68CA6B00AD9C56 /* WKSLinearMediaPlayer.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		A14F9B772B68CA6C00AD9C56 /* LinearMediaPlayer.swift in Sources */ = {isa = PBXBuildFile; fileRef = A14F9B752B68CA6B00AD9C56 /* LinearMediaPlayer.swift */; };
 		A15EEDE61E301CEE000069B0 /* WKPasswordView.h in Headers */ = {isa = PBXBuildFile; fileRef = A15EEDE41E301CEE000069B0 /* WKPasswordView.h */; };
-		A16E66752581A44800EE1749 /* WebKit.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 8DC2EF5B0486A6940098B216 /* WebKit.framework */; };
```
Was this cleanup you did, or was this Xcode?

achristensen07 replied:
Xcode

sysrqb commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
```diff
@@ -0,0 +1,70 @@
+        "    let t = new WebTransport('https://127.0.0.1:%d/');"
+        "    await t.ready;"
+        "    let w = t.datagrams.writable.getWriter();"
+        "    await w.write(new Uint8Array([0x41, 0x42, 0x43]));"
```
It would be nice if these bytes could be echoed back and you could verify it, but I guess that will need to wait until a future PR.

achristensen07 replied:
Yeah, that requires more work.  Getting the server to acknowledge receiving bytes shows that something is happening successfully.  My first attempt at getting it to receive a message didn't work, but it will work soon.

sysrqb reviewed (approved):
`Tools/TestWebKitAPI/cocoa/WebTransportServer.h` and `.mm` seem to be extraneous. But looks like a good start.

achristensen07 committed:
Begin implementing WebTransport network interface
https://bugs.webkit.org/show_bug.cgi?id=278891
rdar://134977323

Reviewed by Matthew Finkel.

As a first step, I start implementing the connection initialization and data sending
using h2 instead of the WebTransport protocol.  They are similar enough that they
both benefit from using the nw_connection_group_t abstraction which is an object
that can be initialized with a URL, start or receive "connections" and send or receive
data.  nw_connection_group_set_receive_handler needs to be called before
nw_connection_group_start so I create a NetworkTransportSession for the receive handler
before deciding if we want to discard it by calling the completion handler of
NetworkTransportSession::initialize with nullptr if there is a network failure during
initialization.  Other than that, this PR basically just adds some straightforward
abstractions and begins implementing a few of them.

WebTransport has some special needs on the server, so I made a new test server for it.
httpd doesn't really support it in a straightforward way, and even WPT's WebTransport
tests require a little custom server setup.  TestWebKitAPI::HTTPServer isn't really
the right abstraction for it, either because I will eventually need to use
nw_listener_set_new_connection_group_handler instead of nw_listener_set_new_connection_handler
to listen for quic connections.  It isn't quite the way it needs to be yet, but
I just listen for one encrypted TCP connection and wait until it receives more than 0 bytes,
which verifies that something is happening on the client.  In my next PR I'll hook up more
things and soon I'll start using nw_connection_group_t on the server too.

* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::messageSenderConnection const):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::createOutgoingUnidirectionalStream):
(WebKit::NetworkTransportSession::createBidirectionalStream):
(WebKit::NetworkTransportSession::receiveDatagram):
(WebKit::NetworkTransportSession::NetworkTransportSession): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm: Added.
(WebKit::NetworkTransportSession::NetworkTransportSession):
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::sendDatagram):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm: Added.
(TestWebKitAPI::enableWebTransport):
(TestWebKitAPI::TEST(WebTransport, Basic)):
* Tools/TestWebKitAPI/WebTransportServer.h: Added.
(TestWebKitAPI::WebTransportServer::bytesReceived const):
(TestWebKitAPI::WebTransportServer::setBytesReceived):
* Tools/TestWebKitAPI/WebTransportServer.mm: Added.
(TestWebKitAPI::connectionLoop):
(TestWebKitAPI::WebTransportServer::WebTransportServer):
(TestWebKitAPI::WebTransportServer::port const):
* Tools/TestWebKitAPI/cocoa/HTTPServer.h:
* Tools/TestWebKitAPI/cocoa/WebTransportServer.h: Added.
* Tools/TestWebKitAPI/cocoa/WebTransportServer.mm: Added.

Canonical link: https://commits.webkit.org/282945@main
